csharp/source/autowrap/csharp/boilerplate.d

```diff
@@ -0,0 +1,164 @@
+/**
+   Necessary boilerplate for C#/.NET.
+
+   To wrap all functions/return/parameter types and struct/class definitions from
+   a list of modules, write this in a "main" module and generate mylib.{so,dll}:
+
+   ------
+   mixin wrapCSharp("mylib", Modules("module1", "module2", ...));
+   ------
+ */
+module autowrap.csharp.boilerplate;
+
+import autowrap.csharp.dlang;
+import autowrap.reflection;
+
+public string wrapCSharp(in string libraryName, in Modules modules) @safe pure {
+    import std.format : format;
+    import std.algorithm: map;
+    import std.array: join;
+
+    if(!__ctfe) return null;
+
+    const modulesList = modules.value.map!(a => a.toString).join(", ");
+
+    return q{
+        //Insert shared boilerplate code
+        %1$s
+
+        import std.typecons;
+        import autowrap.csharp.boilerplate;
+        import autowrap.csharp.dlang : wrapDLang;
+
+        immutable string t = wrapDLang!(%2$s);
+        mixin(t);
+        pragma(msg, t);
+
+        //Insert DllMain for Windows only.
+        version(Windows) {
+            %3$s
+        }
+    }.format(commonBoilerplate(), modulesList, dllMainMixinStr());
+}
+
+/**
+   Returns a string to be mixed in that defines the boilerplate code needed by all C# interfaces.
+ */
+private string commonBoilerplate() @safe pure {
+    return q{
+        import core.memory;
```

Comment:
None of this code can be `@safe` or `pure`?

Reply:
Some of it, maybe, but what would be the point? This is a C interface that is specialized for C#, which doesn't have the notions of (un)safety or purity. Nobody should ever be calling these methods from D code as it will not behave in a manner that one would expect D code to behave and as a result you'll almost certainly make mess. nothrow does have direct implications on return types consumed by C# so I do use it in the generated code as it helps to enforce things at compile-time.